### PR TITLE
Add code folding based on heading level

### PR DESCRIPTION
**Pull request by marieflorescontact**

### CODE REVIEW

File: src/tableOfContentsProvider.ts [hunk 1]
```diff
@@ -28,6 +28,7 @@ export class TableOfContentsProvider {
       try {
         this.toc = await this.buildToc(this.document)
       } catch (e) {
+        console.log(e)
```
Comment (Mogztter) [inline src/tableOfContentsProvider.ts:31]: We should give context, for instance: 

```js
console.log(`Unable to build the Table Of Content for: ${this.document.fileName}`, e)
```

File: src/test/foldingProvider.test.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  test('Should collapse multiple newlines to single newline before next header', async () => {
+    const folds = await getFoldsForDocument(`
+= a
+x
+
+
+
+= b
+y`)
```
Comment (Mogztter) [inline src/test/foldingProvider.test.ts:63]: It should be:

```suggestion
== a
x



== b
y`)
```

We cannot have more than one document title in AsciiDoc.

File: src/test/foldingProvider.test.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    assert.strictEqual(folds.length, 2)
+    const firstFold = folds[0]
```
Comment (Mogztter) [inline src/test/foldingProvider.test.ts:65]: I think we should also add assertion on `folds[1]`

File: src/test/foldingProvider.test.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    assert.strictEqual(firstFold.start, 1)
+    assert.strictEqual(firstFold.end, 5)
+  })
+
+  test('Should not collapse if there is no newline before next header', async () => {
+    const folds = await getFoldsForDocument(`
+= a
+x
+== b
+y`)
```
Comment (Mogztter) [inline src/test/foldingProvider.test.ts:75]: The document title should be on the first line:

```suggestion
    const folds = await getFoldsForDocument(`= a
x
== b
y`)
```

If you apply this change, you will need to update your assertions.

File: src/asciidocParser.ts [hunk 5]
```diff
@@ -91,10 +91,12 @@ export class AsciidocParser {
...
     text: string,
-    doc: vscode.TextDocument,
-    context: vscode.ExtensionContext,
-    editor: vscode.WebviewPanel
-  ): {html: string, document: Asciidoctor.Document} {
+    doc: SkinnyTextDocument,
+    forHTMLSave: vscode.ExtensionContext,
+    backend: vscode.WebviewPanel,
+    context?: vscode.ExtensionContext,
+    editor?: vscode.WebviewPanel
```
Comment (Mogztter) [inline src/asciidocParser.ts:98]: This is incorrect, probably due to the rebase.
It should be:

```suggestion
    context: vscode.ExtensionContext,
    editor: vscode.WebviewPanel
```

Comment (marieflorescontact) [reply src/asciidocParser.ts:98]: it took me a while but I get it now :sweat_smile:

File: src/asciidocParser.ts [hunk 6]
```diff
@@ -91,10 +91,10 @@ export class AsciidocParser {
 
   public convertUsingJavascript (
     text: string,
-    doc: vscode.TextDocument,
+    doc: SkinnyTextDocument,
     context: vscode.ExtensionContext,
     editor: vscode.WebviewPanel
-  ): {html: string, document: Asciidoctor.Document} {
+  ): { html: string; document: Asciidoctor.Document } {
```
Comment (Mogztter) [inline src/asciidocParser.ts:97]: Not sure about the `;` I think it should be `,` no?

**Review by Mogztter — APPROVED** [review]
Perfect 👌🏻

**Pull request merged**